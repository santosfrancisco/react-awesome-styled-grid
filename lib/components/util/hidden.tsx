import { useState, useEffect, Fragment } from "react";
import useEventListener from "@use-it/event-listener";
import { getScreenClass } from "./screens";
import { ConfigProps, HiddenProps } from "../../types";

export const Hidden = ({
  xs,
  sm,
  md,
  lg,
  xl,
  children,
  theme
}: HiddenProps & ConfigProps) => {
  const [currentScreen, setCurrentScreen] = useState("xs");

  const setScreen = () => {
    const newScreen = getScreenClass({ theme });
    if (currentScreen !== newScreen) {
      setCurrentScreen(newScreen);
    }
  };
  useEffect(() => {
    setScreen();
  }, []);

  useEventListener("orientationchange", setScreen);
  useEventListener("resize", setScreen);

  const isHidden = () => {
    switch (currentScreen) {
      case "xs":
        return xs;
      case "sm":
        return sm;
      case "md":
        return md;
      case "lg":
        return lg;
      case "xl":
        return xl;
    }
  };
  if (isHidden()) return false;
  return <Fragment>{children}</Fragment>;
};
