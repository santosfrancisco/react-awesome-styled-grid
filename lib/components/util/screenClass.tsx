import React, { useCallback, useEffect, useState } from "react";
import { getScreenClass } from "./screens";
import { useTheme } from "styled-components";

type ScreenClassProps = {
  render: (screen: string) => React.ReactNode;
};

export const ScreenClass = (props: ScreenClassProps) => {
  const theme = useTheme();
  const [currentScreen, setCurrentScreen] = useState("xs");

  const setScreen = useCallback(() => {
    const lastScreenClass = currentScreen;
    const newScreenClass = getScreenClass({ theme });
    if (!lastScreenClass || lastScreenClass !== newScreenClass) {
      setCurrentScreen(newScreenClass);
    }
  }, []);

  useEffect(() => {
    setScreen();
    if (typeof window !== "undefined") {
      window.addEventListener("orientationchange", setScreen, false);
      window.addEventListener("resize", setScreen, false);
    }
    return () => {
      if (typeof window !== "undefined") {
        window.removeEventListener("orientationchange", setScreen);
        window.removeEventListener("resize", setScreen);
      }
    };
  }, []);

  return <React.Fragment>{props?.render(currentScreen)}</React.Fragment>;
};
