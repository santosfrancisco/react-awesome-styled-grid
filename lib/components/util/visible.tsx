import { useState, useEffect, Fragment } from "react";
import useEventListener from "@use-it/event-listener";
import { ConfigProps, VisibleProps } from "../../types";
import { getScreenClass } from "./screens";

export const Visible = ({
  xs,
  sm,
  md,
  lg,
  xl,
  children,
  theme
}: VisibleProps & ConfigProps) => {
  const [currentScreen, setCurrentScreen] = useState("xs");

  const setScreen = () => {
    const newScreen = getScreenClass({ theme });
    if (currentScreen !== newScreen) {
      setCurrentScreen(newScreen);
    }
  };
  useEffect(() => {
    setScreen();
  }, []);

  useEventListener("orientationchange", setScreen);
  useEventListener("resize", setScreen);

  const isVisible = () => {
    switch (currentScreen) {
      case "xs":
        return xs;
      case "sm":
        return sm;
      case "md":
        return md;
      case "lg":
        return lg;
      case "xl":
        return xl;
    }
  };
  if (!isVisible()) return false;
  return <Fragment>{children}</Fragment>;
};
