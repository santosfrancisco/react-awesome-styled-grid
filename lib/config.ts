import { Config, ConfigProps, GridBreakpoints, Media } from "@site/types";
import { css, CSSObject } from "styled-components";

export const CUSTOM_CONF = "awesomegrid";
export const DIMENSIONS: GridBreakpoints[] = ["xs", "sm", "md", "lg", "xl"];
export const BASE_CONF: Config = {
  media: null,
  mediaQuery: "only screen",
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: "full", // 'full' = max-width: 100%
    sm: "full", // 'full' = max-width: 100%
    md: "full", // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
};

const configs = [];
const hasCustomConf = (props: ConfigProps) => {
  if (!props?.theme) return {};
  return JSON.stringify(props.theme[CUSTOM_CONF] || {});
};
const resolveConfig = (props: ConfigProps) => {
  let themeConf = {};
  if (props?.theme) {
    themeConf = props.theme[CUSTOM_CONF] || {};
  }

  const conf: Config = {
    ...BASE_CONF,
    ...themeConf
  };

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint];
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(" and ")
    );
    return media;
  }, {} as Media);

  return conf;
};

export default function config(props?: ConfigProps): Config {
  const customConf = hasCustomConf(props);
  if (configs[0] === customConf) {
    return configs[1];
  }

  const conf = resolveConfig(props);

  configs[0] = customConf;
  configs[1] = conf;

  return conf;
}

function makeMedia(media: string) {
  return (...args: CSSObject[] | TemplateStringsArray) => css`
    @media ${media} {
      ${css.apply(null, [...args])}
    }
  `;
}
