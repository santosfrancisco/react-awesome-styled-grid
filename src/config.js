import { css } from 'styled-components'

const CUSTOM_CONF = 'awesomegrid'
export const DIMENSIONS = ['xs', 'sm', 'md', 'lg', 'xl']
export const BASE_CONF = {
  mediaQuery: 'only screen',
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: 'full', // 'full' = max-width: 100%
    sm: 'full', // 'full' = max-width: 100%
    md: 'full', // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
}

const configs = []
const hasCustomConf = props => JSON.stringify((props.theme && props.theme[CUSTOM_CONF]) || {})
const resolveConfig = props => {
  const themeConf = (props.theme && props.theme[CUSTOM_CONF]) || {}

  const conf = {
    ...BASE_CONF,
    ...themeConf
  }

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint]
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(' and ')
    )
    return media
  }, {})

  return conf
}

export default function config (props = {}) {
  const customConf = hasCustomConf(props)
  if (configs[0] === customConf) {
    return configs[1]
  }

  const conf = resolveConfig(props)

  configs[0] = customConf
  configs[1] = conf

  return conf
}

function makeMedia (media) {
  return (...args) => css`
    @media ${media} {
      ${css(...args)}
    }
  `
}
