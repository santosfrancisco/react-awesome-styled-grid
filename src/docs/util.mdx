---
name: Utilities
route: /util
order: 2
---

# Utilities

## Media Queries

To use in your css the media queries with the same breakpoints of the grid,
just import the config as in the example below. If you are not using a
custom config, it is not mandatory to pass *props* to config

```js
import React from 'react'
import styled from 'styled-components'
import {Container, Row, Col, config} from 'react-awesome-styled-grid'

export default styled(({ className, children}) => {
  return (
    <Container className={className}>
      <Row>
        <Col>{children}</Col>
      </Row>
    </Container>
  )
})`
  // css here...

  ${props => config(props).media['sm']`
    color: red;
    // ...more css here
  `}

  ${props => config(props).media['md']`
    color: purple;
    font-size: 1.5rem;
    // ...more css here
  `}
`
```

> **Your code**
>
>```js
>${props => config(props).media['sm']`
>  color: red;
>  // ...more css here
>`}
>```
>
>**Your output**
>
>```css
>@media only screen and (min-width: 48rem) {
>  color: red;
>  // ...more css here
>}
>```
