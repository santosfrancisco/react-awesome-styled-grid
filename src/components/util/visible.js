import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { getScreenClass } from '../util/screens'

class Visible extends Component {
  constructor () {
    super()
    this.state = {screen: 'xs'}
    this.setScreen = this.setScreen.bind(this)
  }
  componentDidMount () {
    this.setScreen()
    if (typeof window !== 'undefined') {
      window.addEventListener('orientationchange', this.setScreen, false)
      window.addEventListener('resize', this.setScreen, false)
    }
  }
  setScreen () {
    let lastScreenClass = this.state.screen
    const actualScreenClass = getScreenClass()
    if (!lastScreenClass || (lastScreenClass !== actualScreenClass)) {
      lastScreenClass = actualScreenClass
      this.setState({screen: actualScreenClass})
    }
  }

  componentWillUnmount () {
    window.removeEventListener('orientationchange', this.setScreen)
    window.removeEventListener('resize', this.setScreen)
  }

  isVisible () {
    const { xs, sm, md, lg, xl } = this.props
    switch (this.state.screen) {
      case 'xs':
        return xs
      case 'sm':
        return sm
      case 'md':
        return md
      case 'lg':
        return lg
      case 'xl':
        return xl
      default:
        return false
    }
  }

  render () {
    const { children } = this.props
    return (
      <React.Fragment>
        {this.isVisible() && children}
      </React.Fragment>
    )
  }
}

Visible.displayName = 'Visible'

Visible.propTypes = {
  xs: PropTypes.bool,
  sm: PropTypes.bool,
  md: PropTypes.bool,
  lg: PropTypes.bool,
  xl: PropTypes.bool,
  children: PropTypes.node
}

export default Visible
