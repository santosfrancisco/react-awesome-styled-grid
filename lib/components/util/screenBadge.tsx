import styled, { css } from "styled-components";
import isPropValid from "@emotion/is-prop-valid";
import config, { DIMENSIONS } from "../../config";

export const ScreenBadge = styled.div.withConfig({
  shouldForwardProp: (prop) => isPropValid(prop)
})`
  z-index: 10;
  position: fixed;
  font-size: 1.5rem;
  font-weight: bold;
  right: 10px;
  bottom: 10px;
  width: 50px;
  height: 30px;
  background-color: #5901ad40;
  display: flex;
  border-radius: 5px;
  justify-content: center;
  align-items: center;
  ${(props) => css`
    ${DIMENSIONS.map(
      (dimension) =>
        config(props).breakpoints[dimension] &&
        config(props).media[dimension]`
      &::before {
        content: '${dimension}'
      }
    `
    )}
  `}
`;
