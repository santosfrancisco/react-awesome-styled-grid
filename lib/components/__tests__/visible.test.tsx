import { act, render, screen } from "@testing-library/react";
import config from "../../config";
import { Visible } from "../util/visible";

describe("<Visible />", () => {
  const resize = (width) => {
    global.window.innerWidth = width;
    global.window.dispatchEvent(new Event("resize"));
  };

  it("should render default style correctly", () => {
    const { container } = render(
      <Visible>
        <p>children</p>
      </Visible>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should initialize state", () => {
    act(() => {
      resize(1);
    });
    render(<Visible xs>visible</Visible>);
    expect(screen.queryAllByText("visible")).toHaveLength(1);
    act(() => {
      resize(768);
    });
    expect(screen.queryAllByText("visible")).toHaveLength(0);
  });

  it("should control screen state", () => {
    act(() => {
      resize(1);
    });
    render(<Visible sm>visible</Visible>);
    expect(screen.queryAllByText("visible")).toHaveLength(0);
    act(() => {
      resize(768);
    });
    expect(screen.queryAllByText("visible")).toHaveLength(1);
  });

  it("should always be visible", () => {
    render(
      <Visible xs sm md xl lg>
        visible
      </Visible>
    );
    Object.values(config().breakpoints).forEach((b) => {
      act(() => {
        resize(b * 16);
      });
      expect(screen.queryAllByText("visible")).toHaveLength(1);
    });
  });
});
