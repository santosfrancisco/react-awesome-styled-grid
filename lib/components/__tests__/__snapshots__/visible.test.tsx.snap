// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Visible /> should render default style correctly 1`] = `null`;
